### PR TITLE
feat(js): js sdk preferences

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?
The Preferences module for JS SDK.

### Screenshots

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/client/src/api/api.service.ts [hunk 1]
```diff
@@ -206,14 +207,28 @@ export class ApiService {
...
   }
 
+  async getPreferences({
+    level = PreferenceLevelEnum.TEMPLATE,
+  }: {
+    level?: `${PreferenceLevelEnum}`;
+  }): Promise<Array<IUserPreferenceSettings | IUserGlobalPreferenceSettings>> {
+    return this.httpClient.get(`/widgets/preferences/${level}`);
+  }
+
```
Comment (LetItRock) [inline packages/client/src/api/api.service.ts:231]: We provide newer APIs but it wasn't exposed with this ApiService class.

File: packages/js/src/preferences/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  const level = hasWorkflow ? PreferenceLevel.TEMPLATE : PreferenceLevel.GLOBAL;
+
+  return new Preference({
+    level,
+    enabled: preference.enabled,
+    channels: preference.channels,
+    workflow,
+    overrides: preference.overrides,
+  });
+};
```
Comment (LetItRock) [inline packages/js/src/preferences/helpers.ts]: Map server preference to the one that we use in the module with a better interface.

File: packages/js/src/preferences/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  const { workflowId, enabled, channel } = args;
+  try {
+    emitter.emit('preferences.update.pending', { args });
+
+    let response;
+    if (workflowId) {
+      response = await apiService.updateSubscriberPreference(workflowId, channel, enabled);
+    } else {
+      response = await apiService.updateSubscriberGlobalPreference([{ channelType: channel, enabled }]);
+    }
```
Comment (LetItRock) [inline packages/js/src/preferences/helpers.ts:60]: It seems that we can unify this in the future under one endpoint.

File: packages/js/src/feeds/notification.ts [hunk 4]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
-  subscriber?: Subscriber;
-  transactionId: string;
-  content: string;
-  read: boolean;
-  seen: boolean;
-  deleted: boolean;
-  cta: Cta;
-  payload: Record<string, unknown>;
-  overrides: Record<string, unknown>;
+  readonly _id: string;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]: Let's remove all the fields we don't need on the FE from this Class for now and we will clean them from the API at a later stage.

Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]:
```suggestion
  readonly id: string;
```

File: packages/js/src/feeds/notification.ts [hunk 5]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
+  readonly _feedId?: string | null;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]: We can remove this, notifications will always belong to a feed array.

File: packages/js/src/feeds/notification.ts [hunk 6]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
+  readonly createdAt: string;
+  readonly updatedAt: string;
+  readonly actor?: Actor;
+  readonly subscriber?: Subscriber;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]: Is this necessary for Inbox? If not please remove.

File: packages/js/src/feeds/notification.ts [hunk 7]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
+  readonly transactionId: string;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]: Ditto

File: packages/js/src/feeds/notification.ts [hunk 8]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
+  readonly content: string;
+  readonly read: boolean;
+  readonly seen: boolean;
+  readonly deleted: boolean;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts:24]: Ditto

File: packages/js/src/feeds/notification.ts [hunk 9]
```diff
@@ -35,20 +35,20 @@ export class Notification implements Pick<NovuEventEmitter, 'on' | 'off'> {
...
+  readonly cta: Cta;
+  readonly payload: Record<string, unknown>;
+  readonly overrides: Record<string, unknown>;
```
Comment (SokratisVidros) [inline packages/js/src/feeds/notification.ts]: Ditto.

File: packages/js/src/preferences/helpers.ts (hunk 3 above)
Comment (SokratisVidros) [reply packages/js/src/preferences/helpers.ts:60]: Correct. Let's add a comment in the code about that.

File: packages/js/src/preferences/preference.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import { ApiService } from '@novu/client';
+
+import { NovuEventEmitter } from '../event-emitter';
+import { ChannelPreference, ChannelPreferenceOverride, ChannelType, PreferenceLevel, WorkflowInfo } from '../types';
+import { ApiServiceSingleton } from '../utils/api-service-signleton';
+import { updatePreference } from './helpers';
```
Comment (SokratisVidros) [inline packages/js/src/preferences/preference.ts:6]: Nit-picky suggestion: Move the helpers into the Preference class to have everything in one place and and avoid an extra import. Every byte counts ;)

File: packages/js/src/preferences/preferences.ts [hunk 11]
```diff
@@ -1,3 +1,32 @@
 import { BaseModule } from '../base-module';
```
Comment (SokratisVidros) [inline packages/js/src/preferences/preferences.ts:1]: For a future PR: The naming is a bit off. Is this a `BaseModule` or `BaseResource`, or `BaseModel`?

File: packages/js/src/types.ts [hunk 12]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
   action?: MessageAction;
 };
 
+export type WorkflowInfo = {
```
Comment (SokratisVidros) [inline packages/js/src/types.ts]:
```suggestion
export type Workflow = {
```

File: packages/js/src/types.ts [hunk 13]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
...
+  _id: string;
+  name: string;
+  critical: boolean;
+  tags?: string[];
```
Comment (SokratisVidros) [inline packages/js/src/types.ts]: Do we need this at this stage? If no, please remove.

File: packages/js/src/types.ts [hunk 14]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
...
+  identifier: string;
```
Comment (SokratisVidros) [inline packages/js/src/types.ts:125]: What's the difference between an `identifier` and the `_id`?

File: packages/js/src/types.ts [hunk 15]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
   action?: MessageAction;
 };
 
+export type WorkflowInfo = {
+  _id: string;
```
Comment (SokratisVidros) [inline packages/js/src/types.ts]:
```suggestion
  id: string;
```

File: packages/js/src/types.ts [hunk 16]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
...
+  tags?: string[];
+  identifier: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  data?: Record<string, any>;
+};
+
+export type ChannelPreference = {
+  email?: boolean;
+  sms?: boolean;
+  in_app?: boolean;
```
Comment (SokratisVidros) [inline packages/js/src/types.ts:133]:
```suggestion
  inApp?: boolean;
```

File: packages/js/src/types.ts [hunk 17]
```diff
@@ -99,6 +118,29 @@ export type Cta = {
...
+  chat?: boolean;
+  push?: boolean;
+};
+
+export type ChannelPreferenceOverride = {
```
Comment (SokratisVidros) [inline packages/js/src/types.ts]: This seems overly complicated for the current frontend needs and is driven by the current state of the API. Let's build the DX on an ideal scenario and hide the API tech-debt with internal helper methods. Then we will update the API.

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: packages/js/src/feeds/notification.ts (hunk 5 above)
Comment (LetItRock) [reply packages/js/src/feeds/notification.ts]: but you might want to know to which feed it belongs when using individually

File: packages/js/src/preferences/helpers.ts (hunk 3 above)
Comment (LetItRock) [reply packages/js/src/preferences/helpers.ts:60]: https://linear.app/novu/issue/COM-60/consolidate-the-update-preferences-global-template-under-one-endpoint

File: packages/js/src/preferences/preference.ts (hunk 10 above)
Comment (LetItRock) [reply packages/js/src/preferences/preference.ts:6]: can't do that as it's reused in two places: single Preference class and Preferences module

File: packages/js/src/preferences/preferences.ts (hunk 11 above)
Comment (LetItRock) [reply packages/js/src/preferences/preferences.ts:1]: yes I was thinking what could be a better name here, what you would suggest instead?

File: packages/js/src/types.ts (hunk 14 above)
Comment (LetItRock) [reply packages/js/src/types.ts:125]: `identifier` is used for the trigger, `_id` is internal Mongo id

File: packages/js/src/types.ts (hunk 16 above)
Comment (LetItRock) [reply packages/js/src/types.ts:133]: this is due to internal implementation and will require changes across the stack :(

File: packages/js/src/types.ts (hunk 17 above)
Comment (LetItRock) [reply packages/js/src/types.ts]: I think that we can remove it from the new Preference class for now, not even sure if that is used often by clients.

File: packages/js/src/event-emitter/novu-event-emitter.ts [hunk 18]
```diff
@@ -1,23 +1,34 @@
...
   }
 
   on<Key extends EventNames>(eventName: Key, listener: EventHandler<Events[Key]>): void {
-    this.emitter.on(eventName, listener);
+    this.#mittEmitter.on(eventName, listener);
   }
 
   off<Key extends EventNames>(eventName: Key, listener: EventHandler<Events[Key]>): void {
-    this.emitter.on(eventName, listener);
+    this.#mittEmitter.on(eventName, listener);
```
Comment (BiswaViraj) [inline packages/js/src/event-emitter/novu-event-emitter.ts]:
```suggestion
    this.#mittEmitter.off(eventName, listener);
```

Comment (BiswaViraj) [reply packages/js/src/event-emitter/novu-event-emitter.ts]: Sorry, noticed it is already fixed in the other PR 👌🏼

**Review by BiswaViraj — APPROVED** [review]

**Pull request merged**